Fixup: correct the ECIES for the P-521 curve as per SEC 1

vkhoroz opened the pull request:
It appears that the go-ethereum and its derivatives only ever tested the P-256 and S-256 curves for ECIES. The tests for other curves are missing (apart from the params check). What's worse, the P-521 curve shared key generation panics with the current parameters, and also fail when unmarshalling the public key of the encrypted message inside Decrypt.

So, I made 2 commits to make the P-521 curve work correctly. See the commit messages for more context of each change.

The changed `GenerateShared` function still pass the test vectors in https://github.com/foundriesio/go-ecies/pull/4 which were generated using the previous implementation (including the P-521 curve). The reason why they pass for P-521 is because old GenerateShared tests ignored curve parameters and used a max key length. Ironically, tests did it the correct way; it's the `Encrypt/Decrypt` code which was passing wrong parameters.

Likewise, the `Decrypt` function passes the test vectors in https://github.com/foundriesio/go-ecies/pull/4 which were generated after fixing the `GenerateShared` function, but before fixing the `Decrypt` function. Actully, these tests fail for the P-521 curve without the second commit due to wrong unmarshalling in the original implementation.

Signed-off-by: Volodymyr Khoroz <volodymyr.khoroz@foundries.io>

vkhoroz commented:
@doanac I was going to comment out the support for the P-521 curve.

But then I thought: hey it is doing the wrong thing for the P-256 either; it's just a coincidence that it produced the correct result.

This PR also helps to resolve one more mistery I did not understand:
https://github.com/foundriesio/fioconfig/blob/main/internal/ecies_pkcs11.go#L86

The PKCS11 don't accept the parameters go-ethereum added in (synthetically), and still produced the same key for P-256 curve.
So, now I have the answer: the keylength was equal to maxkeylength for P-256.

Given that the end result for P-256 is the same; and the P-521 curve now doesn't panic... I think we can still keep it.
I'm actually glad that we have a curve which is not a power of 2... as it helps to get rid of apparently wrong code.

vkhoroz commented:
Original error: https://github.com/ethereum/go-ethereum/blob/master/crypto/ecies/ecies.go#L136
```
copy(sk[len(sk)-len(skBytes):], skBytes)
```

For the P-521 curve `len(sk) == 64` and `len(skBytes) == 66` which results into an index out of range panic.

vkhoroz commented:
@ldts I think we need your cryptography expertise to verify these changes.

ldts commented on ecies.go:
```diff
@@ -263,16 +263,20 @@ func (prv *PrivateKey) Decrypt(rand io.Reader, c, s1, s2 []byte) (m []byte, err
 		err = ErrInvalidPublicKey
 		return
 	}
+	if len(c) < (rLen + hLen + 1) {
```
I would code it a bit different: use `kLen  = prv.PublicKey.Curve.Params().BitSize+7)/8` then instead of rLen I would use `mStart = 1 + kLen` in the non-compressed format and `mStart = 1 + 2 * kLen` in the compressed format...

ldts replied:
other than that looks good to me (I thikn line 276 the +1 is **not** required btw)

vkhoroz replied:
I wanted the change to be minimalistic... but reusing `kLen  = prv.PublicKey.Curve.Params().BitSize+7)/8` makes sense to me - I will do that.

The `rLen` is used down in the below code, so semantically it makes sense to have both `mStart` and `rLen`.

Also at line 276 I guess `+ 1` is to accommodate for at least 1 character of the encrypted message itself. I'm not sure if the empty message is a good edge case to support. So will leave that as is in the original too.

vkhoroz replied:
@ldts updated in force-push.

I actually found a way to part with rLen easily while keeping the code semantics intact.

vkhoroz commented:
I think it is worth to mention that the changes to the `GenerateShared` make it compatible with the https://github.com/libtom/libtomcrypt/blob/v1.18.2/src/pk/ecc/ecc_shared_secret.c#L32 too, not only https://github.com/golang/go/blob/go1.19.5/src/crypto/tls/key_schedule.go#L167 and (unreleased) https://github.com/golang/go/blob/master/src/crypto/ecdh/nist.go#L191. So, we now have at least two authoritative sources which prove that this change is correct.

Thanks to @ldts for mentioning that.

ldts reviewed (approved):
LGTM

vkhoroz commented:
@ldts sorry, I somehow messed the commit history, now it is fixed.

I added the zeropad fix we discussed offline.
Quite surprizingly, it was easy to generate keys where this was breaking for the P-521 curve. I even had quite a few messages in [encryption vector tests](https://github.com/foundriesio/go-ecies/pull/4) generating wrong output.

Knowing the problem it didn't take too much time to find a handful of such keys for the P-384 curve and even 1 key for the P-256 curve. I added these findings to the test vectors (as can be seen in fixups here: https://github.com/foundriesio/go-ecies/pull/4).
